sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala

```diff
@@ -225,10 +228,25 @@ case class CachedRDDBuilder(
     _cachedColumnBuffers
   }
 
+  def manualClose[T <: CachedBatch](entry: T): T = {
+    entry match {
+      case o: AutoCloseable =>
+        try {
+          o.close
+        } catch {
+          case NonFatal(e) =>
+            logWarning("Fail to close a memory entry", e)
+        }
+      case _ =>
+    }
+    entry
+  }
+
   def clearCache(blocking: Boolean = false): Unit = {
     if (_cachedColumnBuffers != null) {
       synchronized {
         if (_cachedColumnBuffers != null) {
+          _cachedColumnBuffers.foreach(manualClose)
           _cachedColumnBuffers.unpersist(blocking)
```

Comment:
Sorry I may miss some context. Doesn't `_cachedColumnBuffers.unpersist(blocking)` ask the memory store to free the memory entry and release the memory?

Reply:
@cloud-fan , hmm, from our test, since _cachedColumnBuffers is RDD[CachedBatch], after _cachedColumnBuffers.unpersist(blocking) was called, the reference of all CachedBatch will be deleted, just like what happened in MemoryStore.

Reply:
`RDD.unpersist` will go to the new code path you added in `MemoryStore`, and release the memory, or do I miss something?

Reply:
Did you set `blocking` to true in your test?

Reply:
> `RDD.unpersist` will go to the new code path you added in `MemoryStore`, and release the memory, or do I miss something?

Oh, I didn't realize that, I thought cachedBatch required an extra close if it hasn't been serialized/deserialized to persist... I just verified of what you said, and cachedBatch did get released without calling _cachedColumnBuffers.foreach(manualClose).

Thanks for pointing out! Anyway, in that case, do you think it will be helpful if I keep the UT to make sure user-defined cachedBatch is closed? or I will close this PR.

Reply:
It's always good to add more test cases!